open-notes/maturity-model.md

```diff
@@ -0,0 +1,394 @@
+
+### Keep current
+
+Not bleeding edge. We upgrade packages to the latest SemVer minor version as they are released. For frameworks like .NET we use the LTS version and trail by one minor version.
+
+### Command Query Separation
+
+https://www.martinfowler.com/bliki/CommandQuerySeparation.html
+
+### Dependency aware health checks
+
+Distinguish in your health checks between a service being healthy, its dependencies for partial availability, and completely unavailable. For example if an applications database is unavailable, it cannot do anything. If on the other hand a service where it gets fresh data from that it then caches is unavailable, it can still function although with stale data.
+
+https://github.com/beamly/SE4  
+https://hootsuite.github.io/health-checks-api/
+
+### Retro all the things
+
+Agile practices is about learning often and folding that learning into how you behave. Scrum is a starting point but is really just a minimum bar. By taking a moment to retro every standup, every meeting, every whiteboard session, and every pair-programming session, we set ourselves up to improve at an exponentially more rapid pace than when we restrict our reflections to once a sprint. Start with standup and retro it every day. See what happens.  
+https://devonburriss.me/agile-is-a-characteristic/
+
+### Deploy daily
+
+To drive quality up by automating and and learning rapidly, we will deploy any application that changes every day. This is still assuming some sort of manual intervention of sign-off but all intermediate steps like deployment and testing are already automated to make this just a gate.
+
+### Honest types
+
+Known as Value types in DDD, these are a replacement of simple types like `string` and `int` with more meaningful and smarter types. See https://devonburriss.me/honest-return-types/
+
+### Pair programming
+
+A pair of developers work together on a single story. One is the "driver" and is the one controlling the keyboard. The other is the navigator and is thinking through and suggesting what to do.
+
+https://www.davefarley.net/?p=267
+
+### Correctness alerts
+
+Using your business metrics, monitor what is normal and alert when that normal is not met. As an example if you are monitoring the conversion rate in a step and that drops, that can indicate a problem that is not necessarily throwing exceptions but is effecting customers.
+
+### APM tracing
+
+Instrument your code so it is possible to visualize and measure important calls within and between services.
+
+This article mentions the main standards:  
+https://www.datadoghq.com/blog/instrument-opencensus-opentracing-and-openmetrics/
+
+### Resilient requests
+
+Requests between services need to acknowledge that the network is not reliable. Requests should cater for this by supporting retries. If the dependency is down, a decision needs to be made about partial availability and continuation later or rollback of changes so far.
+
+###  Embrace immutable code
```

Comment:
The language is different with this skill. Less directive. Personally, I like the suggestion type (embrace), rather than directive.